Jetpack Search/Sync: ensure syncing of post meta and taxonomies that are indexed

gibrown opened the pull request:
Our search index includes a fixed list of post meta keys and custom taxonomies in order to limit the number of fields in the index (see documentation https://jetpack.com/support/search/frequently-asked-questions/#troubleshoot-custom-tax).

We have so far not been syncing the list from wp.com into Jetpack which has had many downsides:
- Uncertainty if it worked in all cases (also lack of tests)
- No easy way to add new fields
- No visibility into what is indexed
- No way to double check that the config that the user has will work.
- No way to provide post meta faceting in search

This fixes all of that.

The class-search.php file will be replacing the old-style class.jetpack-sync-module-search.php that we are currently using on wp.com for the indexing. See https://github.com/Automattic/wpes-lib/blob/master/src/indices/jetpack-search-v2-index.php#L842

Fixes #15672
Fixes #8663

#### Does this pull request change what data or activity we track or use?

This potentially expands the list of post meta that we sync a lot. I don't think that is a real change to what we need to document though.

#### Testing instructions:

I tested this with a new set of unit tests. The unit tests are one of the more important pieces since I expect that the list of post meta and taxonomies will evolve over time and we want to make sure that it is trivial to add a new field.

Warning: this adds 1608 tests to the sync testsuite. They take almost 5 minutes to run. I've already spent a bit of time trying to reduce the time though (I cut out doing any update/delete tests). I think checking every item is worth it unfortunately. I found a number of keys that just don't work or that are weird (hello post_format taxonomy!).

```
yarn docker:phpunit --filter=WP_Test_Jetpack_Sync_Search
yarn docker:phpunit --testsuite=sync
```

#### Proposed changelog entry for your changes:
* Search: adds explicit post meta keys and taxonomies to sync into the index. Particularly important for integrating with third-party plugins.

gibrown committed:
Jetpack Search: ensure syncing of post meta and taxonomies in the index

gibrown committed:
Jetpack Search: ensure syncing of post meta and taxonomies in the index

gibrown committed:
Remove copy paste mistake.

gibrown committed:
Remove copy paste mistake.

gibrown commented on tools/phpcs-excludelist.json:
```diff
@@ -538,6 +538,7 @@
 	"projects/plugins/jetpack/tests/php/sync/test_class.jetpack-sync-posts.php",
 	"projects/plugins/jetpack/tests/php/sync/test_class.jetpack-sync-queue.php",
 	"projects/plugins/jetpack/tests/php/sync/test_class.jetpack-sync-sender.php",
+	"projects/plugins/jetpack/tests/php/sync/test_class.jetpack-sync-search.php",
```
This doesn't seem to be preventing the linter from failing for some reason.

jeherve replied:
I'm afraid not. You would still need to have code that is free of linting errors for all new and modified lines in the file (so the whole file in this case).

gibrown committed:
Remove unnecessary code.

gibrown committed:
Remove unnecessary code.

robertf4 committed:
Recommendations: Adjust banner styles (#18849)

Co-authored-by: Jeff Golenski <jeffgolenski@gmail.com>

blackjackkent committed:
Add Cloudflare Analytics snippet to site header via (WPCOM only) Jetpack plugin (#18771)

* Add Cloudflare Analytics snippet to site header via (WPCOM only) Jetpack plugin

* Updates based on PR feedback

* Ensure the feature is also enabled on wpcom simple sites

See https://github.com/Automattic/jetpack/pull/18771#issuecomment-777033437

* Update option name.

Co-authored-by: Jeremy Herve <jeremy@jeremy.hu>

ebinnion committed:
SSO: Disregard filters to use in-place connection when redirecting after SSO (#18801)

tjcafferkey committed:
Add advanced menu items for General and Writing for WP Admin (#18804)

mmtr committed:
Admin Menu: Fix link destination for Widgets and Menus (#18862)

robertf4 committed:
Setup Wizard: Remove setup wizard (#18867)

Co-authored-by: Jeremy Herve <jeremy@jeremy.hu>

jeherve committed:
Actions: switch from app to action for Stale issue/PR labeling (#18859)

Co-authored-by: Brad Jorsch <anomiex@users.noreply.github.com>

jeherve committed:
Instagram: use arguments provided by shortcode attributes (#18768)

jsnmoon committed:
Instant Search: Fix theming issues for the new modal design (#18852)

sergeymitr committed:
Connection: updating the deprecated `is_user_connected()` method (#18876)

The method Jetpack::is_user_connected() is deprecated by PR #18548.
Replacing it with its counterpart from the jetpack-connection package.

pablinos committed:
Correct the strokeWidth attribute. (#18845)

gibrown committed:
Clean up the linting errors for the tests.

gibrown committed:
Clean up the linting errors for the tests.

gibrown committed:
Move checking of synced meta test from Search to Meta test class.

gibrown committed:
Move checking of synced meta test from Search to Meta test class.

gibrown committed:
Remove assertIsArray()

gibrown committed:
Remove assertIsArray()

gibrown committed:
assertIsArray does not work in php 5.6

gibrown committed:
assertIsArray does not work in php 5.6

gibrown committed:
assertIsBool also does not exist in 5.6

gibrown committed:
assertIsBool also does not exist in 5.6

gibrown commented:
I'm not sure what is up with the remaining failures.

jeherve committed:
VideoPress: avoid warning when no rating is set (#18857)

In some scenarios, the following notice happens:
PHP Notice:  Undefined property: stdClass::$rating in /home/holden/apps/jeherve_com/w/wp-content/plugins/jetpack-dev/_inc/lib/core-api/wpcom-fields/class-wpcom-rest-api-v2-attachment-videopress-data.php on line 97

Let's make sure we always have at least a default rating.

jeherve committed:
Actions: make milestone detection available for all tasks (#18558)

Co-authored-by: Brad Jorsch <anomiex@users.noreply.github.com>

jeherve committed:
Actions: do not add milestone info to comment when undefined (#18881)

Follow-up from #18558.

When we do not have any plugin info, let's not try to add anything to the comment.

jeherve committed:
Actions: do not run label and comment automations on closed PRs (#18882)

Follow-up to #18558.

We do not have to keep editing labels or comments on PRs that are closed.

andyjdavis committed:
Avoid unnecessary converting of booleans to strings in the WordAds module (#18823)

mmtr committed:
Keyring: Register Sharing page without menu (#18861)

mdbitz committed:
Update Checksums to support blacklisted taxonomies. (#18779)

* Update Checksums to support blacklisted taxonomies.
new structured helper functions in Settings
join field settings to support id and range differences
remove range query from parent join

* Remove unnecessary operator `<>` when it could be replaced with `NOT IN`

* Remove the parent table range exceptions and replace them with nulls on the call

Co-authored-by: Biser Perchinkov <biser@bisko.be>

anomiex committed:
CLI: Fix CodeQL warning (#18877)

We tried to match the regex documented at
https://getcomposer.org/doc/04-schema.md#name for allowed package names.
But somehow we lost the `/` in the middle. We really only wanted the
part after the slash: `^[a-z0-9](([_.]?|-{0,2})[a-z0-9]+)*$`

But that wouldn't satisfy CodeQL. We could simplify that to the more
efficient `^[a-z0-9]+(([_.]|-{1,2})[a-z0-9]+)*$`

But that's not what Composer is actually using (or, AFAICT, has ever
used). They just use `^[a-z0-9_.-]+$`, see
https://github.com/composer/composer/blob/03e8cacd/src/Composer/Command/InitCommand.php#L239

jeherve committed:
Issues templates: update label name for Beta plugin (#18884)

pablinos committed:
Podcast API Endpoint: Add the publish date to the track information. (#18855)

* Add the publish date to the track information.

* Update the documentation and ensure null for unknown values.

* Update the date format to Atom

* Make the publish_date API doc more specific

* Add the format for validation of the publish_date

Co-authored-by: Grant Kinney <creativecoder@users.noreply.github.com>

Co-authored-by: Grant Kinney <creativecoder@users.noreply.github.com>

leogermani committed:
skip master user tests when in userless mode (#18887)

robertf4 committed:
Recommendations: Hide JITMs when banner is displaying (#18853)

gravityrail committed:
Refactor secrets and tokens (#18812)

glendaviesnz committed:
Set up testing-library/react for testing jetpack blocks and add example tests to WhatsApp block (#18582)

Co-authored-by: Aaron Robertshaw <60436221+aaronrobertshaw@users.noreply.github.com>
Co-authored-by: Glen Davies <glen.davies@a8c.com>

sgomes committed:
Remove jQuery from Related Posts (#18874)

Co-authored-by: Jeremy Herve <jeremy@jeremy.hu>

glendaviesnz committed:
Carousel: Improve handling of touch device events (#18720)

Co-authored-by: Glen Davies <glen.davies@a8c.com>

mmtr committed:
Fix Publicize Services endpoint on WP.com (#18886)

leogermani committed:
Adds message to Instagram widget form for disconnected users (#18891)

frontdevde committed:
Update Color Studio and make use of it in Admin Color Schemes (#18873)

* Update color studio dependency to 2.4.0

* Update Aquatic color scheme to draw on color studio for color values

* Update Classic Blue color scheme to draw on color studio for color values

* Update Classic Bright color scheme to draw on color studio for color values

* Update Classic Dark color scheme to draw on color studio for relevant color values

* Update Contrast color scheme to draw on color studio for color values

* Update Nightfall color scheme to draw on color studio for color values

* Update Powder Snow color scheme to draw on color studio for color values

* Update Sakura color scheme to draw on color studio for color values

* Update Sunset color scheme to draw on color studio for color values

* Update Sunset color scheme to draw on color studio for color values II

* Update comments and values where applicable

mmtr committed:
Admin Menu: Add "Advanced Users Management" menu on WoA sites (#18805)

leogermani committed:
Update/instagram widget userless (#18902)

* improve Instagram widget form for disconn. users

* instagram widget: only show message in JP plugin

mmtr committed:
Admin Menu: Exclude hidden items from API response (#18898)

jeherve committed:
Recommendations: update button in My Plan to go to Recommendations (#18843)

Commit:
Ads: Fixes check for displaying the Ads module settings (#18893)

mdbitz commented on projects/packages/sync/src/modules/class-search.php:
```diff
@@ -0,0 +1,1793 @@
+	 */
+	public function __construct() {
+		// Post meta whitelists.
+		add_filter( 'jetpack_sync_post_meta_whitelist', array( $this, 'add_search_post_meta_whitelist' ), 10 );
```
I'd suggest we gate this addition to only when Instant Search is active. This has the potential to increase the number of actions flowing to WP.com and we should be responsible to only send if the site is using Instant Search.

mdbitz replied:
also we'd need a test on this conditional as well.

Search not active -> these meta are not synced
Search active -> these meta are synced.

--

Yes this will lead to concerns that when users purchase Instant Search the Cache site doesn't have all necessary data, however that already is a concern and we trigger a Full Sync to resolve. This will be further mitigated by rollout of Fixes and so I recommend we maintain not syncing these fields unless Instant Search is active/enabled.

mdbitz commented on projects/packages/sync/src/modules/class-search.php:
```diff
@@ -0,0 +1,1793 @@
+	 *
+	 * @var array
+	 */
+	private static $taxonomies_to_sync = array(
```
@gibrown -> can you outline what the purpose of this is?  In Sync the taxonomies sent to WP.com are determined by a blacklist [$blacklisted_taxonomies](https://github.com/Automattic/jetpack/blob/7d0fcdfc99c9fc2a1d1ddf60858c0a292cbdf385/projects/packages/sync/src/class-defaults.php#L415). Why are we maintaining a manual list of taxonomies that don't control actual data flows?

Ooo I get it this isn't for Jetpack's usage but instead is a whitelist on the WP.com side when creating the elastic doc object.  

If that is correct I might suggest a unit test that no values of $taxonomies_to_sync are contained in $blacklisted_taxonomies

mdbitz reviewed (changes requested):
@gibrown :wave: Overall this looks good, I added several notes around testing and the only "blocker" I see is that we don't want the meta added unless Instant Search is enabled as there is a possibility of spiking the amount of content being processed which wouldn't be utilized.

mdbitz commented:
@gibrown Sorry another follow-up question :) Can you outline what testing steps you performed on the WP.com side to verify these new meta where persisted to the cache sites?

gibrown commented:
Trying to reproduce e2e tests ended up completely breaking my docker environment so I'm not going to be able to rush this out. I think for the current release that the woocommerce sync stuff will work for us fine anyways, so I'll work on cleaning up the points from this review and getting it merged later this week. We can always cherry pick if we find a reason to but I doubt we will.

gibrown committed:
Mostly working fixes to feedback that needs a bit more review and testing.

[not verified]

gibrown committed:
Merge branch 'add/jp-search-sync' of github.com:Automattic/jetpack into add/jp-search-sync

bluefuton commented:
@gibrown I think the rebase might have gone a bit wonky?

bluefuton commented:
Closing in favour of https://github.com/Automattic/jetpack/pull/19674.
